### ISSUE TITLE
Enable Module Federation (Upgrade to Webpack 5)

### ISSUE THREAD

**Issue by pmuir** [strimzi/strimzi-ui#54]
We intend to use [Webpack Module Federation](https://webpack.js.org/concepts/module-federation/) to build Red Hat product UIs from Strimzi UI.

In order to do this we will need to enable Module Federation in Strimzi UI, which will involve:

* Add this to the relevant architecture documentation
* Upgrading Webpack to the latest v5 release (Module Federation was added in Webpack 5)
* Upgrading all Webpack loaders & plugins to the latest version (this will ensure they are compatible with Webpack 5)
* Add the module federation plugin to the webpack config, and declare source files we want to expose
* Add shared dependencies (`react`, `react-dom`) as shared dependencies
* Add default exports to each source file we want to expose (with the default export pointing at the component we are exporting)

This also has an impact on how we design our React components as we need to start designing them for reuse in unknown situations, not just use within in the strimzi-ui. Designing components for reuse via module federation is no different to designing them for reuse via npm dependency management.

A few component design rules that will make reuse much easier are:

* create "pages" that wrap components. For each view that we want to create, build the component separately from the page, allowing the the component to be reused. A page might have titles, padding, layout etc. The component is the actual interaction.
* create "dumb" components that are wrapped by "smart" components (this is also very useful for mocking data for testing, storybook etc.). Here for each component we create we pass in any data that is needed using props, rather than directly accessing it. This allows the data sources to be altered in various ways by the reusing UI. Smart components are often pages as well.
* use callbacks to handle transitions to/from views in dumb components. In other words don't navigate using `history` but instead pass a callback in as a prop. This allows the component to be reused in a different setting.
* make components more granular. Don't build the entire view in a single component - split it up to allow reuse.

**Comment by matthew-chirgwin** [comment]
I am all for keeping dependencies, particulalry key ones such as Webpack, React etc, up to date, having delt with the cost in tech debt of not in the past. If there are features we can then use to make the UI even more able, thats even better.

I happen to have done some high level reading around Module Federation, but not done anything with it in anger - its a neat (and scary!) feature. While I don't think it will add/enhancethe experience of a user of the Strimzi-ui, it will enforce the behaviours we have detailed in component approach/implementaion (keeping things presentational, pushing page context up, keep components detatched/dumb to make the federation possible). Thus, I am definatley open to the idea of using it here.

I would be interested to see/understand how 'in anger':

- The code changes needed to the build and component code to make use of Module Federation
  - How this effects the multiple view layer capability of the UI
- The resulting built output (for UI/Server)

My main concerns at this point are:

- We want to get to a topics listing page ASAP to validate the foundation/allow the addition of new capabilities
  - This for me needs to happen before we make significant changes - and act as a validation of this new build/bundle approach
- We are still letting things like TypeScript (and required configuration) settle - I worry if we change things in the near term we may break/miss something

**Comment by pmuir** [comment]
Currently storybook is not able to work with Webpack 5 - see https://github.com/storybookjs/storybook/issues/9216. There seems to be work going on currently to try to fix it. Will put this on hold for now.
